cpp/include/cudf/join_new.hpp

```diff
@@ -0,0 +1,345 @@
+                                   Filter filter,
+                                   rmm::mr::device_memory_resource* mr = rmm::mr::get_default_resource());
+
+  template <typename Filter>
+  std::unique_ptr<table> full_join(cudf::table_view const& left,
+                                   cudf::table_view const& right,
+                                   std::vector<std::pair<cudf::size_type, cudf::size_type>> const& columns_in_common,
+                                   std::vector<cudf::size_type> const& base_left_on,
+                                   std::vector<cudf::size_type> const& base_right_on,
+                                   Filter filter,
+                                   rmm::mr::device_memory_resource* mr = rmm::mr::get_default_resource());
+} //namespace sort_merge
+
+namespace hash {
+  template <typename Filter>
+  std::unique_ptr<table> inner_join(cudf::table_view const& left,
+                                    cudf::table_view const& right,
+                                    std::vector<std::pair<cudf::size_type, cudf::size_type>> const& columns_in_common,
+                                    std::vector<cudf::size_type> const& base_left_on,
+                                    std::vector<cudf::size_type> const& base_right_on,
+                                    Filter filter,
+                                    rmm::mr::device_memory_resource* mr = rmm::mr::get_default_resource());
+
+  template <typename Filter>
+  std::unique_ptr<table> left_join(cudf::table_view const& left,
+                                   cudf::table_view const& right,
+                                   std::vector<std::pair<cudf::size_type, cudf::size_type>> const& columns_in_common,
+                                   std::vector<cudf::size_type> const& base_left_on,
+                                   std::vector<cudf::size_type> const& base_right_on,
+                                   Filter filter,
+                                   rmm::mr::device_memory_resource* mr = rmm::mr::get_default_resource());
+
+  template <typename Filter>
+  std::unique_ptr<table> full_join(cudf::table_view const& left,
+                                   cudf::table_view const& right,
+                                   std::vector<std::pair<cudf::size_type, cudf::size_type>> const& columns_in_common,
+                                   std::vector<cudf::size_type> const& base_left_on,
+                                   std::vector<cudf::size_type> const& base_right_on,
+                                   Filter filter,
+                                   rmm::mr::device_memory_resource* mr = rmm::mr::get_default_resource());
+} //namespace hash
+
+//
+//  Option 3:  Something completely different
+//
+namespace completely_different {
+
+  /**
+   * @brief Type defining join comparison operators
+   *
+   * Indicates how two elements `a` and `b` compare with one and another.
+   */
+  enum class join_comparison_operator {
```

Comment:
This enum allows us to define the non equi join clauses as well.  We can start by implementing just EQUAL (add CUDF_EXPECTS calls to throw if some comparison other than EQUAL is specified).